### PR TITLE
bpo-31344: Per-frame control of trace events

### PR DESCRIPTION
**Pull request by ncoghlan**
f_trace_lines: enable/disable line trace events
f_trace_opcodes: enable/disable opcode trace events

These are intended primarily for testing of the interpreter
itself, as they make it much easier to emulate signals
arriving at unfortunate times.

https://bugs.python.org/issue31344

### CODE REVIEW

**Comment by zooba** [comment]
Ping @int19h - any thoughts? :)

**Review by zooba — APPROVED** [review]
Looks good to me (with a post-8pm caveat)

**Comment by int19h** [comment]
This would potentially come in handy for debuggers; e.g. we could implement per-statement (rather than per-line) stepping with this.

File: Doc/reference/datamodel.rst [hunk 1]
```diff
@@ -970,10 +970,20 @@ Internal types
...
 
       Special writable attributes: :attr:`f_trace`, if not ``None``, is a function
-      called at the start of each source code line (this is used by the debugger);
+      called for various events during code execution (this is used by the debugger).
+      Normally an event is triggered for each new source line - this can be
+      disabled by setting :attr:`f_trace_lines` to :const:`False`.
+
+      Implementations *may* allow per-opcode events to be requested by setting
+      :attr:`f_trace_opcodes` to :const:`True`. This is intended specifically
+      for interpreter signal safety testing, and may have surprising results
```
Comment (gpshead) [inline Doc/reference/datamodel.rst]: perhaps change the wording to:

_and may result in surprising undefined behavior if an exception escapes the trace function._

i'm trying to make it more explicit that we do not guarantee any behavior when exceptions are allowed to escape during opcode tracing.  changing "are raised by" to something involving "escape" makes it clear that exceptions can be raised and handled.  the important thing is that they are resolved before the function exits.

File: Include/pystate.h [hunk 2]
```diff
@@ -92,14 +92,19 @@ typedef struct _is {
 /* Py_tracefunc return -1 when raising an exception, or 0 for success. */
 typedef int (*Py_tracefunc)(PyObject *, struct _frame *, int, PyObject *);
 
-/* The following values are used for 'what' for tracefunc functions: */
+/* The following values are used for 'what' for tracefunc functions
+ *
+ * To add a new kind of trace event, also update "trace_init" in
```
Comment (gpshead) [inline Include/pystate.h:97]: I'm gonna guess you learned this the hard way.  thanks for the comment! :)

File: Lib/test/test_sys_settrace.py [hunk 3]
```diff
@@ -234,12 +234,21 @@ def generator_example():
 
 
 class Tracer:
-    def __init__(self):
+    def __init__(self, trace_line_events=None, trace_opcode_events=None):
+        self.trace_line_events = trace_line_events
+        self.trace_opcode_events = trace_opcode_events
         self.events = []
+    def _reconfigure_frame(self, frame):
```
Comment (gpshead) [inline Lib/test/test_sys_settrace.py]: style nit, leave a blank line between methods

File: Lib/test/test_sys_settrace.py [hunk 4]
```diff
@@ -257,6 +266,10 @@ def tearDown(self):
         if self.using_gc:
             gc.enable()
 
+    @staticmethod
+    def make_tracer():
```
Comment (gpshead) [inline Lib/test/test_sys_settrace.py:274]: add a comment mentioning why is this needed at all. (letting subclasses such as XYZ override it)

File: Lib/test/test_sys_settrace.py [hunk 5]
```diff
@@ -398,6 +411,33 @@ def func():
...
+
+
+@support.cpython_only
+class TraceOpcodesTestCase(TraceTestCase):
+    """Repeat the trace tests, but with per-opcodes events enabled"""
+
+    def compare_events(self, line_offset, events, expected_events):
+        skip_opcode_events = [e for e in events if e[1] != 'opcode']
+        if len(events) > 1:
+            self.assertLess(len(skip_opcode_events), len(events))
```
Comment (gpshead) [inline Lib/test/test_sys_settrace.py]: add a msg= parameter describing why this is a problem if this were to fail.

File: Doc/reference/datamodel.rst [hunk 6]
```diff
@@ -970,10 +970,20 @@ Internal types
...
          single: f_lineno (frame attribute)
 
       Special writable attributes: :attr:`f_trace`, if not ``None``, is a function
-      called at the start of each source code line (this is used by the debugger);
+      called for various events during code execution (this is used by the debugger).
+      Normally an event is triggered for each new source line - this can be
+      disabled by setting :attr:`f_trace_lines` to :const:`False`.
+
+      Implementations *may* allow per-opcode events to be requested by setting
+      :attr:`f_trace_opcodes` to :const:`True`. This is intended specifically
```
Comment (gpshead) [inline Doc/reference/datamodel.rst]: Lets not describe what it is intended for here.  that we happen to use it for interpreter signal exception safety testing isn't relevant to someone reading about the API.  It'll gain other uses.

**Review by gpshead — COMMENTED** [review]
overall good, some comments to address.

**Review by gpshead — APPROVED** [review]

**Comment by ncoghlan** [comment]
Merging as the previous Appveyor run already passed: https://ci.appveyor.com/project/python/cpython/build/3.7.0a0.6126
